### PR TITLE
feat: support custom components for OpenAPI generation

### PR DESCRIPTION
**Pull request by aarnphm**
## What does this PR address?

Address #2142 #2570

- feat: adding openapi specification attrs class
- feat: serde with attrs representation for OpenAPI
- chore: adding/fixing documentations for IODescriptor.

This PR ensures OpenAPI generation are matching specs. This PR will also address
supports for components schemas.

TODO:
- Pydantic Exception model https://pydantic-docs.helpmanual.io/usage/validators/

## Before submitting:

- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [x] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [x] Did you write tests to cover your changes?

### CODE REVIEW

File: bentoml/_internal/service/openapi/specification.py [hunk 1]
```diff
@@ -0,0 +1,545 @@
+from __future__ import annotations
```
Comment (aarnphm) [inline bentoml/_internal/service/openapi/specification.py:1]: Source of truth: https://spec.openapis.org/oas/v3.1.0#openapi-specification

File: bentoml/_internal/io_descriptors/base.py [hunk 2]
```diff
@@ -5,83 +5,81 @@
...
+
+    def __new__(cls: t.Type[Self], *args: t.Any, **kwargs: t.Any) -> Self:
         self = super().__new__(cls)
-        arg_strs = tuple(repr(i) for i in args) + tuple(
-            f"{k}={repr(v)}" for k, v in kwargs.items()
-        )
-        setattr(self, "_init_str", f"{cls.__name__}({', '.join(arg_strs)})")
+        # default mime type is application/json
+        self._mime_type = "application/json"
+        self._init_str = cls.__qualname__
```
Comment (aarnphm) [inline bentoml/_internal/io_descriptors/base.py:50]: reason behind this change is that the inferenceAPI name doesn't look nice on OpenAPI
Screenshot 2022-08-02 at 20 25 26

File: bentoml/_internal/io_descriptors/numpy.py [hunk 3]
```diff
@@ -140,39 +148,55 @@ def __init__(
...
+    def openapi_responses(self) -> OpenAPIResponse:
+        example = (
+            self._input_sample.tolist() if self._input_sample is not None else None
+        )
+
+        return OpenAPIResponse(
+            description=SUCCESS_DESCRIPTION,
+            content={
+                self._mime_type: MediaType(
+                    schema=self.openapi_schema(), example=example
```
Comment (aarnphm) [inline bentoml/_internal/io_descriptors/numpy.py:196]: added examples if users use `from_sample`

File: bentoml/_internal/io_descriptors/json.py [hunk 4]
```diff
@@ -139,41 +138,66 @@ def classify(input_data: IrisFeatures) -> np.ndarray:
...
+        # returns schemas from pydantic_model.
+        return Schema(
+            **schema.model_process_schema(
+                self._pydantic_model,
+                model_name_map=schema.get_model_name_map(
+                    schema.get_flat_models_from_model(self._pydantic_model)
+                ),
+                ref_prefix=REF_PREFIX,
+            )[0]
+        )
```
Comment (aarnphm) [inline bentoml/_internal/io_descriptors/json.py:180]: support pydantic models components now :)

**Review by aarnphm — COMMENTED** [review]
Changes from IO descriptor fits with the new OpenAPI schemas.

TODO: 
- generate examples. 
- add tests for pandas dataframe/series.

File: bentoml/_internal/service/openapi/__init__.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+merger = Merger(
+    # merge dicts
+    [(dict, "merge")],
+    # override all other types
+    ["override"],
+    # override conflicting types
+    ["override"],
+)
+
```
Comment (aarnphm) [inline bentoml/_internal/service/openapi/__init__.py:56]: We want to merge repeated components here.

File: docs/source/conf.py [hunk 6]
```diff
@@ -66,8 +66,11 @@
 myst_enable_extensions = ["colon_fence"]
 
 # Remove the prompt when copying examples
-copybutton_prompt_text = r">>> |\.\.\.|» |$ |% "
+copybutton_prompt_text = r">>> |\.\.\.|» |\% |\$ |In \[\d*\]: | {2,5}\.\.\.: | {5,8}: "
```
Comment (aarnphm) [inline docs/source/conf.py:69]: handles comments:
```
asdfa
# some comments
```
 -> copies will only return `asdfa`

**Review by bojiang — APPROVED** [review]

**Pull request merged**